pkg/k8s_helper/k8s_helper.go

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: MIT
+// SPDX-FileCopyrightText: (c) Copyright 2023 Advanced Micro Devices, Inc.
+package k8s_helper
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type KubernetesHelper struct {
+	Config *rest.Config
+}
+
+// Create the new helper with the in-cluster config.
+func NewKubernetesHelper() (*KubernetesHelper, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return &KubernetesHelper{
+		Config: config,
+	}, nil
+}
+
+type ContainerHelper interface {
+	GetContainerID(pod *corev1.Pod) (*string, error)
+}
+
+type DefaultContainerHelper struct {
+	ContainerName string
+}
+
+func (h *DefaultContainerHelper) GetContainerID(pod *corev1.Pod) (*string, error) {
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Name != h.ContainerName {
+			continue
+		}
+
+		if status.Ready && *status.Started && status.State.Running != nil {
+			return &status.ContainerID, nil
+		}
+
+		return nil, nil
+	}
+
+	return nil, nil
+}
+
+func NewDefaultContainerHelper(containerName string) ContainerHelper {
+	return &DefaultContainerHelper{
+		ContainerName: containerName,
+	}
+}
+
+// Get ID of the container identified by the pod's name, namespace
+// and also the container's name if there are more than one in the pod.
+// The returned string is in the format "<type>://<container_id>".
+func (k8s *KubernetesHelper) GetContainerID(ctx context.Context,
+	podNamespace string, podName string, helper ContainerHelper,
+) (string, error) {
+	cs, err := kubernetes.NewForConfig(k8s.Config)
+	if err != nil {
+		return "", err
+	}
+
+	for {
+		pod, err := cs.CoreV1().Pods(podNamespace).Get(
+			ctx, podName, metav1.GetOptions{})
+		if err != nil {
+			return "", err
+		}
+
+		containerID, err := helper.GetContainerID(pod)
+		if err != nil {
+			return "", err
+		}
+
+		if containerID != nil {
+			return *containerID, nil
+		}
+
+		// "Backoff" and sleep with context.
+		select {
+		case <-time.After(1 * time.Second):
+			continue
+		case <-ctx.Done():
+			return "", errors.New("Timeout")
+		}
```

Comment:
Thanks, the `flowcontrol` package from the `client-go` repo is new to me. 👍  I tried to find backing-off packages before coding, but everything I saw would be a new third-party dependency, which I wanted to avoid.

There is not much of a documentation for it, so I jotted down a quick demo which I'll embed into the PR shortly:

```diff
diff --git a/cmd/worker/main.go b/cmd/worker/main.go
index cc4e7de..4275aa8 100644
--- a/cmd/worker/main.go
+++ b/cmd/worker/main.go
@@ -12,8 +12,25 @@ import (

        "github.com/Xilinx-CNS/kubernetes-onload/pkg/control_plane"
        "github.com/Xilinx-CNS/kubernetes-onload/pkg/k8s_helper"
+
+       "k8s.io/client-go/util/flowcontrol"
 )

+func testBackoff() {
+       b := flowcontrol.NewBackOff(100*time.Millisecond, 1*time.Second)
+
+       for i := 0; i < 16; i++ {
+               b.Next("foo", time.Now())
+
+               glog.Infof("%d", i)
+
+               select {
+               case <-time.After(b.Get("foo")):
+                       continue
+               }
+       }
+}
+
 func main() {
        // Enable logging to stderr.
        flag.Parse()
@@ -22,6 +39,8 @@ func main() {
                glog.Fatalf("Failed to initialise Onload worker: %v", err)
        }

+       testBackoff()
+
        // Get the container identification from env.
        podNamespace := os.Getenv("POD_NAMESPACE")
        podName := os.Getenv("POD_NAME")
```

Produces:
```console
$ oc logs -c onload-worker onload-sample-onload-device-plugin-ds-9vh79
I0907 10:58:11.127920 1403805 main.go:25] 0
I0907 10:58:11.228218 1403805 main.go:25] 1
I0907 10:58:11.428754 1403805 main.go:25] 2
I0907 10:58:11.829732 1403805 main.go:25] 3
I0907 10:58:12.630423 1403805 main.go:25] 4
I0907 10:58:13.631359 1403805 main.go:25] 5
I0907 10:58:14.631557 1403805 main.go:25] 6
I0907 10:58:15.631830 1403805 main.go:25] 7
I0907 10:58:16.632052 1403805 main.go:25] 8
I0907 10:58:17.632210 1403805 main.go:25] 9
I0907 10:58:18.633356 1403805 main.go:25] 10
I0907 10:58:19.633842 1403805 main.go:25] 11
I0907 10:58:20.633912 1403805 main.go:25] 12
I0907 10:58:21.634043 1403805 main.go:25] 13
I0907 10:58:22.634181 1403805 main.go:25] 14
I0907 10:58:23.634264 1403805 main.go:25] 15
I0907 10:58:24.649791 1403805 main.go:76] Found container ID cri-o://9bd22046b19a4df43428f2ad493af1ceb22b7d447e9f1a65d5dbc1be9c79a970
I0907 10:58:24.649818 1403805 control_plane.go:23] CRI-O container ID is 9bd22046b19a4df43428f2ad493af1ceb22b7d447e9f1a65d5dbc1be9c79a970
I0907 10:58:24.650251 1403805 control_plane.go:32] Updated Onload control plane server path to /usr/bin/crictl
I0907 10:58:24.650524 1403805 control_plane.go:41] Updated Onload control plane server params to exec 9bd22046b19a4df43428f2ad493af1ceb22b7d447e9f1a65d5dbc1be9c79a970 /mnt/onload/sbin/onload_cp_server -K
```

There is the exponentially growing delay with the upper limit.

Reply:
Done.